nebula_decoders/include/nebula_decoders/nebula_decoders_velodyne/decoders/velodyne_generic_decoder.hpp

```diff
@@ -0,0 +1,445 @@
+  {
+    const raw_packet_t * raw = (const raw_packet_t *)&velodyne_packet.data[0];
+    double last_azimuth_diff = 0;
+    uint16_t azimuth_next;
+    const uint8_t return_mode = velodyne_packet.data[RETURN_MODE_INDEX];
+    const bool dual_return = (return_mode == RETURN_MODE_DUAL);
+
+    int num_padding_blocks = sensor_.getNumPaddingBlocks(dual_return);
+
+    for (uint block = 0; block < static_cast<uint>(BLOCKS_PER_PACKET - num_padding_blocks);
+         block++) {
+      // Cache block for use.
+      const raw_block_t & current_block = raw->blocks[block];
+
+      uint bank_origin = 0;
+      // Used to detect which bank of 32 lasers is in this block.
+      switch (current_block.header) {
+        case BANK_1:
+          bank_origin = 0;
+          break;
+        case BANK_2:
+          bank_origin = 32;
+          break;
+        case BANK_3:
+          bank_origin = 64;
+          break;
+        case BANK_4:
+          bank_origin = 96;
+          break;
+        default:
+          // Do not flood the log with messages, only issue at most one
+          // of these warnings per minute.
+          return;  // bad packet: skip the rest
+      }
+
+      double azimuth_diff;
+      uint16_t azimuth;
+
+      // Calculate difference between current and next block's azimuth angle.
+      if (block == 0) {
+        azimuth = current_block.rotation;
+      } else {
+        azimuth = azimuth_next;
+      }
+      if (block < static_cast<uint>(BLOCKS_PER_PACKET - (1 + dual_return))) {
+        // Get the next block rotation to calculate how far we rotate between blocks
+        azimuth_next = raw->blocks[block + (1 + dual_return)].rotation;
+
+        // Finds the difference between two successive blocks
+        azimuth_diff = static_cast<double>((36000 + azimuth_next - azimuth) % 36000);
+
+        // This is used when the last block is next to predict rotation amount
+        last_azimuth_diff = azimuth_diff;
+      } else {
+        // This makes the assumption the difference between the last block and the next packet is
+        // the same as the last to the second to last. Assumes RPM doesn't change much between
+        // blocks.
+        azimuth_diff = (block == static_cast<uint>(BLOCKS_PER_PACKET - (4 * dual_return) - 1))
+                         ? 0
+                         : last_azimuth_diff;
+      }
+
+      // Condition added to avoid calculating points which are not in the interesting defined area
+      // (cloud_min_angle < area < cloud_max_angle).
+      if (
+        (sensor_configuration_->cloud_min_angle < sensor_configuration_->cloud_max_angle &&
+         azimuth >= sensor_configuration_->cloud_min_angle * 100 &&
+         azimuth <= sensor_configuration_->cloud_max_angle * 100) ||
+        (sensor_configuration_->cloud_min_angle > sensor_configuration_->cloud_max_angle &&
+         (azimuth <= sensor_configuration_->cloud_max_angle * 100 ||
+          azimuth >= sensor_configuration_->cloud_min_angle * 100))) {
```

Comment:
For all if-statements that filter out points, please convert them to guard clauses:
Instead of
```c++
if (conditionA) {
  // ...
  if (conditionB) {
    // decode point
  }
}
```
invert the conditions and skip the point, resulting in much less nested code:
```c++
if (!conditionA) {
  continue;
}
// ...
if (!conditionB) {
  continue;
}
// decode point
```